### ISSUE TITLE
@mustCallSuper is satisfied by super calls in if statements

### ISSUE THREAD

**Issue by shyndman** [dart-lang/sdk#41357]
Hello,

It appears that the `@mustCallSuper` annotation is satisfied by a super call appearing in a single branch of the subclass' method. This does not trigger a lint: 

```dart
class BaseClass {
  @mustCallSuper
  void foo() {}
}

class SubClass extends BaseClass {
  @override
  void foo() {
    if (Random().nextBool()) {
      super.foo();
    }
  }
}
```

My expectation would be an exhaustive check across the method's branches. 

There are a bunch of [existing bugs](https://github.com/dart-lang/sdk/issues?q=mustCallSuper) open around this annotation, but none appeared to match what I describe (after a brief scan).

```
Flutter 1.16.2 • channel unknown • unknown source
Framework • revision 58cad787d8 (2 weeks ago) • 2020-03-23 08:33:41 -0700
Engine • revision 2d42c74a34
Tools • Dart 2.8.0 (build 2.8.0-dev.16.0 ba8baa46b4)
```

**Comment by zelamngo** [comment]
https://groups.google.com/a/google.com/g/dart-discuss/c/78r1XcxxRwI
https://github.com/dart-lang/sdk/issues/35372
Related.

**Comment by jamesderlin** [comment]
I don't see how that's related.

Regarding this issue: one reason to prefer the existing behavior is to allow early exits from a derived method:

```dart
class Base {
  @mustCallSuper
  void doSomeOperation() {
    if (conditionA) return;
    if (conditionB) return;
    // ... Do actual work...
  }
}

class Derived extends Base {
  @override
  void doSomeOperation() {
    if (conditionC) return;
    super.doSomeOperation();
  }
}
```
I suppose that arguably such a situation might warrant adding `// ignore: must_call_super`, although it appears that that must be applied to the entire function and not on individual return paths.

**Comment by srawlins** [comment]
@pq do you know if this was fixed by your recent work on mustCallSuper?

**Comment by pq** [comment]
No.  I didn't add any flow analysis and I think that's what we'd need to handle this more thoughtfully.

**Comment by srawlins** [comment]
Ah thats right. Will be tricky

**Comment by srawlins** [comment]
Here's my proposal for changing the `@mustCallSuper` warning to use some flow analysis:

* We track in an overridding method body where the code _throws_ (via an explicit `throw`, `rethrow`, or `Never`-typed expression), _returns_ (via an explicit `return` statement), and _calls super_.
* The goal is to require that an overriding method body _definitely calls super or throws._

So let's look at a few cases. Assume `super.m` is annotated with `@mustCallSuper`.

```dart
// Warning!
void m() {
  if (1 == 2) super.m();
}
```

```dart
// Warning!
void m() {
  if (1 == 2) return;
  super.m();
}
```


```dart
// OK!
void m() {
  if (1 == 2) {
    super.m();
  } else if (2 == 3) {
    super.m();
  } else {
    super.m();
  }
}
```

These are the motivating example in the issue summary. Let's look at some loops:

```dart
// Warning!
void m(List<int> list) {
  for (var i in list) {
    super.m(list);
  }
}
```

```dart
// Warning!
void m(bool b) {
  while (b) {
    super.m(b);
    // ...
  }
}
```

```dart
// OK!
void m(bool b) {
  do {
    super.m(b);
    // ...
  } while (b);
}
```

We are not guaranteed to enter the loop body in the first two cases, but we are in the third. I can't think of an example where you call `super` in a loop, so I don't think this will come up much.

What about throwing?

```dart
// OK!
void m(String s) {
  if (!validValues.contains(s)) {
    throw ArgumentError.value(s);
    // OR
    fail(""); // if `fail` returns `Never`.
  }
  super.m(s);
}
```

It makes sense to allow throwing as the one control flow case that does not require calling super. I think in most cases, when an exception needs to be thrown, it is non-sensical to call `super`. You can then think of the annotation as "overriding members must call super in every normal (non-throwing) control flow case." As a matter of fact, since many expressions can throw without making it implicit, this sort of matches the expectation for other cases:

```dart
// OK!
void m(String? s) {
  super.m(s?.toUpperCase());
  // OR
  super.m(s!);
  // etc
}
```

That leads to another proposal that diverges from regular flow analysis: try/catch/finally. Let's look at an example:

```dart
// OK!
void m(String s) {
  try {
    // ...
    super.m(something(s));
  } on X {
    // ...
  } finally {
    // ...
  }
```

In strict flow analysis, we cannot say that `super.m` is called by the end of this method, because `// ...` could throw. Even `something(s)`, which is evaluated before the super call, could throw! However, I think it would be very counter-productive to warn on this code. As long as the block of code in the `try` "definitely calls super or throws," it's fine. We're not concerned with variables maybe not being initialized in a try-body, or variables maybe being promoted in a try-body.

I have a WIP CL supporting this behavior; I hope to mail soon.

**Comment by srawlins** [comment]
I implemented the plan above in a complete CL: https://dart-review.googlesource.com/c/sdk/+/469760, but I've encountered a number of real-world cases where method overrides no longer satisfy the mustCallSuper requirements.

For example, take Flutter's [`CanvasKitRenderer.initialize()`](https://github.com/flutter/flutter/blob/master/engine/src/flutter/lib/web_ui/lib/src/engine/canvaskit/renderer.dart#L50): this method calls `super.initialize()`, but only if `_initialized` hasn't been set. The angular source code has a lot of similar cases: In a mustCallSuper override, we first check some cache, and then call `super` only if necessary.

Then we have [`_CupertinoScrollbarState.handleTrackTapDown`](https://github.com/flutter/flutter/blob/77dba65d63fcfd3aa4f22cb87d5d2793c0b46957/packages/flutter/lib/src/cupertino/scrollbar.dart#L219), which very much just calls super in one path, but not another.

With these examples, we can see that we might want to continue this new strictness in mustCallSuper, and fix existing code. Fixing existing code would be changing an override to always call super (seems not applicable in these cases), or `// ignore: must_call_super` (seems fine to me), or remove the `@mustCallSuper` annotation on the super class.

If we want to have our cake and eat it too though, we could make the annotation configurable, perhaps with a `strict: true|false` parameter. `@MustCallSuper(strict: true|false)`, and then `@mustCallSuper` is an alias either for `strict: true` or `strict: false`.

I don't have a strong opinion here, but I think I'd prefer that, going forward, `@mustCallSuper` means `strict: true`. In any of these cases, where an override _sometimes_ calls super, we can look into changing the annotation on the super method to `@MustCallSuper(strict: false)`. In `strict: false` mode, we just enforce the same behavior today: the overriding method calls `super` _somewhere_, at least in one place.
